File: Retos/Reto #20 - LA TRIFUERZA [Media]/ruby/test0n3.rb
# frozen_string_literal: true

# class TriForce
class TriForce
  # def initialize(input)
  #   @nro_row = input
  #   @padding = (@nro_row * 4 - 1) / 2
  #   @iter = 1
  #   @triangle = ''
  # end

  # def draw_triangles
  #   @nro_row.times do
  #     print_heads
  #     print_tails
  #   end
  #   @triangle
  # end

  # def print_heads
  #   @triangle += ' ' * @padding
  #   @iter.times do
  #     @triangle += '*   '
  #   end
  #   @triangle += "\n"
  #   @padding -= 1
  # end

  # def print_tails
  #   @triangle += ' ' * @padding
  #   @iter.times do
  #     @triangle += '*** '
  #   end
  #   @triangle += "\n"
  #   @iter += 1
  #   @padding -= 1
  # end

  def initialize(input)
    @nro_rows = input
    @element_width = 2 * @nro_rows - 1
    @max_length = @element_width * 2 + 1
    @external_padding = @max_length / 2
    @triforce = ''
  end

  def draw_triforce
    print_head
    print_base
    @triforce
  end

  def print_head
    @nro_rows.times do |i|
      @triforce += "#{' ' * @external_padding}#{gen_stars(i)}\n"
      @external_padding -= 1
    end
  end

  def print_base
    @nro_rows.times do |i|
      @triforce += "#{' ' * @external_padding}#{gen_stars(i)}#{' ' * @element_width}#{gen_stars(i)}\n"
      @external_padding -= 1
      @element_width -= 2
    end
  end

  private

  def gen_stars(row)
    '*' * (row * 2 + 1)
  end
end

# puts TriForce.new(1).draw_triforce
# puts TriForce.new(2).draw_triforce
# puts TriForce.new(3).draw_triforce
# puts TriForce.new(4).draw_triforce
# puts TriForce.new(5).draw_triforce

File: Retos/Reto #20 - LA TRIFUERZA [Media]/javascript/test0n3.js
function Triforce(input) {
  let nroRows = input;
  let elementWidth = 2 * nroRows - 1;
  let maxLength = elementWidth * 2 + 1;
  let externalPadding = Math.floor(maxLength / 2);
  let triforceText = drawTriforce();

  function printHead() {
    let element = "";
    for (let i = 0; i < nroRows; i++) {
      element += `${createPadding(externalPadding)}${createStars(i)}\n`;
      externalPadding -= 1;
    }
    return element;
  }

  function printBase() {
    let element = "";
    for (let i = 0; i < nroRows; i++) {
      element += `${createPadding(externalPadding)}${createStars(
        i
      )}${createPadding(elementWidth)}${createStars(i)}\n`;
      externalPadding -= 1;
      elementWidth -= 2;
    }
    return element;
  }

  function createPadding(padding) {
    let temp = "";
    for (let i = 0; i < padding; i++) {
      temp += " ";
    }
    return temp;
  }

  function createStars(row) {
    let temp = "";
    for (let i = 0; i < row * 2 + 1; i++) {
      temp += "*";
    }
    return temp;
  }

  function drawTriforce() {
    return printHead() + printBase();
  }

  Object.defineProperties(this, {
    triforce: {
      get: function () {
        return triforceText;
      },
    },
    nroRows: {
      get: function () {
        return nroRows;
      },
    },
  });
}

// [1, 2, 5].forEach((i) => console.log(new Triforce(i).triforce));

// console.log(new Triforce(10).nroRows);
